libfdcore/peers.c

```diff
@@ -197,10 +200,16 @@ int fd_peer_remove ( DiamId_t diamid, size_t diamidlen )
 			peer->p_hdr.info.config.pic_flags.exp = PI_EXP_INACTIVE;
 			peer->p_hdr.info.config.pic_lft = 0;
 			CHECK_FCT( fd_p_expi_update(peer) );
+			found = 1;
 			break;
 		}
 	}
 
+	if (!found)
+	{
+		TRACE_ERROR("Diameter peer %*s not valid - caller is out of sync", (int)diamidlen, diamid);
```

Comment:
I appreciate there’s no error from an external point of view, which is why this is a TRC rather than and ENT log say (I think it falls into a similar category to the “Unexpected host” log at the end of fd_peer_hook_cb – i.e. an unexpected code path that might be a symptom of a wider problem).  The reason why I thought we should have something here is that, while testing various HSS connection failure/config error scenarios I was thrown by the fact that the freeDiameter and RealmManager/DiameterStack peers lists seemed to be mutually out of sync.  Eventually I discovered that this was because of freeDiameter’s habit of zombifying peers, but I would have got to that conclusion much sooner if I could be sure that there were no deletes that freeDiameter wasn’t expecting.

Given that FD apparently zombifies peers when they go away, I would have thought its unlikely for an explicit delete from the stack to cross over with a delete from FD.  If you think it is likely, could we change it to an INFO log (so that it at least stands out from DEBUG logs)?

From: Andy Caldwell [mailto:notifications@github.com]
Sent: 10 September 2015 17:08
To: Metaswitch/freeDiameter
Cc: Steve Lloyd
Subject: Re: [freeDiameter] Reviewer AMC: Add/remove peer diagnostics (#26)


In libfdcore/peers.c<https://github.com/Metaswitch/freeDiameter/pull/26#discussion_r39178626>:

>                      break;

>              }

>       }

>

> +     if (!found)

> +     {

> +            TRACE_ERROR("Diameter peer %*s not valid - caller is out of sync", (int)diamidlen, diamid);

I think this is probably a valid path to fall in to (if a peer disappears while we're deleting it). Also, errors on deleting a non-existent object are often misleading (there's no "error" here from an external point of view) and I generally avoid them.

The log at the beginning of the function gets a :+1 from me though!

—
Reply to this email directly or view it on GitHub<https://github.com/Metaswitch/freeDiameter/pull/26/files#r39178626>.